### PR TITLE
Memory-optimized ResizeKernelMap

### PR DESCRIPTION
**Pull request by antonfirsov**
### Prerequisites
- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
In most cases Resize kernel maps are periodic, and it's possible to represent them in a [compact form](https://gist.github.com/antonfirsov/dad7cd0d07ae4885b0c7518df6ab6ec4#file-optimizedkernelmap-md) instead of the [original brute force form](https://gist.github.com/antonfirsov/dad7cd0d07ae4885b0c7518df6ab6ec4#file-oldkernelmap-md). This means that eg. for a `1280 -> 300` resize kernel map it is sufficient to store `15 + 2 * Radius` rows instead of `300 + 2 * Radius` rows.

I tried to make the code explanatory, but finding the rules was mostly about solving math riddles & equations, so it's hard to explain the details without a whiteboard.

### Further details
- Renamed `KernelMap` to `ResizeKernelMap`
- Added a bunch of test cases which are testing the new `ResizeKernelMap` code against a simplified reference implementation
- Switched  to `double` in `ResizeKernelMap` calculations for better accuracy
- Introduced a utility called `TolerantMath` to deal with numeric robustness issues
- Fixed the backing buffer, so we can store pointers in `ResizeKernel` (`Memory<float>` is just too heavy-weight for those small slices of values). (-> I expect some speedup from this, but haven't becnhmarked yet.)
- Moved resamplers to a separate (non-namespace-providing) folder
- Moved Resize-related classes to a separate (non-namespace-providing) folder

### CODE REVIEW

**Comment by antonfirsov** [comment]
@JimBobSquarePants I also have executed the benchmarks and the results are very similar (almost no  speedup for Core, significant speedup for classic Clr).
I had to apply huge hacks on my system due to live audio latency issues, and it behaves somewhat slower now for high CPU load benchmarks, making my old results no longer applicable as baseline.

File: src/ImageSharp/Processing/Processors/Transforms/Resize/ResizeKernelMap.cs [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+            if (length > this.data.Width)
+            {
+                throw new InvalidOperationException(
+                    $"Error in KernelMap.CreateKernel({dataRowIndex},{left},{right}): left > this.data.Width");
+            }
+
+            Span<float> rowSpan = this.data.GetRowSpan(dataRowIndex);
+
+            ref float rowReference = ref MemoryMarshal.GetReference(rowSpan);
+            float* rowPtr = (float*)Unsafe.AsPointer(ref rowReference);
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/Processors/Transforms/Resize/ResizeKernelMap.cs]: Ah yes, missed that. Didn't see the `Pin()` method on `Memory<T>`.

**Comment by antonfirsov** [comment]
@JimBobSquarePants managed to improve my comments a bit.

Actually, the main reason behind this PR is that I realized I can AVX2-optimize the convolution. For that we need to widen all kernel values from `float` to `Vector4`. With the original implementation the kernel map would grow too big in memory. With this optimization it should be tolerable for the majority of the cases. (Except when `sourceLength` or `destinationLength` is a large prime, having no common divisor, but I think this is usually not true for real life cases.)

**Comment by JimBobSquarePants** [comment]
> Actually, the main reason behind this PR is that I realized I can AVX2-optimize the convolution. For that we need to widen all kernel values from float to Vector4.

@antonfirsov I'll have to take your word for it. I can't see that optimization myself.

**Comment by antonfirsov** [comment]
I'm merging this, so I can push the next PR from my queue. Thanks for the review!

**Pull request merged**